### PR TITLE
Handle nil response from backend for diagnostics

### LINKED ISSUE #26640: Neovim doesn't seem to pull diagnostics from Ruby-LSP
### Problem

When I purposely create an error in a simple Ruby file (`puts "Hello, World!`), I notice that Neovim will initially report a diagnostic error message. If I fix the error, I notice the diagnostic message is still present.

The [Ruby-LSP](https://github.com/Shopify/ruby-lsp) is somewhat unique in that it only supports pull diagnostics and not PublishDiagnostics. Although I'm aware the former was implemented in #24128.

### Steps to reproduce using "nvim -u minimal_init.lua"

Steps to reproduce:

1. Ensure you have Ruby 3.0 or greater installed
2. Install Ruby-LSP with `gem install ruby-lsp` (current version 0.13.2)
3. Open Neovim with the following command: `nvim -u [path to minimal_init.lua file] example.rb`
4. Once in the file, invalidate the code in the `example.rb` file by removing one of the quotation marks from `puts "Hello, World"`. You'll see a diagnostic error
5. Put the quotation mark back. You'll still see the error.

Files outlined below:

minimal_init.lua file:

```lua
local on_windows = vim.loop.os_uname().version:match("Windows")

local function join_paths(...)
  local path_sep = on_windows and "\\" or "/"
  local result = table.concat({ ... }, path_sep)
  return result
end

vim.cmd([[set runtimepath=$VIMRUNTIME]])

local temp_dir = vim.loop.os_getenv("TEMP") or "/tmp"

vim.cmd("set packpath=" .. join_paths(temp_dir, "nvim", "site"))

local package_root = join_paths(temp_dir, "nvim", "site", "pack")
local lspconfig_path = join_paths(package_root, "test", "start", "nvim-lspconfig")

if vim.fn.isdirectory(lspconfig_path) ~= 1 then
  vim.fn.system({ "git", "clone", "https://github.com/neovim/nvim-lspconfig", lspconfig_path })
end

vim.lsp.set_log_level("debug")
require("vim.lsp.log").set_format_func(vim.inspect)

local nvim_lsp = require("lspconfig")
local on_attach = function(_, bufnr)
  local function buf_set_option(...)
    vim.api.nvim_buf_set_option(bufnr, ...)
  end

  buf_set_option("omnifunc", "v:lua.vim.lsp.omnifunc")

  -- Mappings.
  local opts = { buffer = bufnr, noremap = true, silent = true }
  vim.keymap.set("n", "gD", vim.lsp.buf.declaration, opts)
  vim.keymap.set("n", "gd", vim.lsp.buf.definition, opts)
  vim.keymap.set("n", "K", vim.lsp.buf.hover, opts)
  vim.keymap.set("n", "gi", vim.lsp.buf.implementation, opts)
  vim.keymap.set("n", "<C-k>", vim.lsp.buf.signature_help, opts)
  vim.keymap.set("n", "<space>wa", vim.lsp.buf.add_workspace_folder, opts)
  vim.keymap.set("n", "<space>wr", vim.lsp.buf.remove_workspace_folder, opts)
  vim.keymap.set("n", "<space>wl", function()
    print(vim.inspect(vim.lsp.buf.list_workspace_folders()))
  end, opts)
  vim.keymap.set("n", "<space>D", vim.lsp.buf.type_definition, opts)
  vim.keymap.set("n", "<space>rn", vim.lsp.buf.rename, opts)
  vim.keymap.set("n", "gr", vim.lsp.buf.references, opts)
  vim.keymap.set("n", "<space>e", vim.diagnostic.open_float, opts)
  vim.keymap.set("n", "[d", vim.diagnostic.goto_prev, opts)
  vim.keymap.set("n", "]d", vim.diagnostic.goto_next, opts)
  vim.keymap.set("n", "<space>q", vim.diagnostic.setloclist, opts)
end

-- Add the server that troubles you here
local name = "ruby_ls"
local cmd = { "ruby-lsp" } -- needed for elixirls, lua_ls, omnisharp
if not name then
  print("You have not defined a server name, please edit minimal_init.lua")
end
if not nvim_lsp[name].document_config.default_config.cmd and not cmd then
  print([[You have not defined a server default cmd for a server
    that requires it please edit minimal_init.lua]])
end

nvim_lsp[name].setup({
  cmd = cmd,
  on_attach = on_attach,
})
```

example.rb file:

```ruby
puts "Hello, World!"
```

### Expected behavior

Diagnostic messages to be continually updated when the Ruby file is edited.

### Neovim version (nvim -v)

0.10.0-dev-7d279a0

### Language server name/version

ruby-lsp 0.13.2

### Operating system/version

macOS 14.1 (23B74)

### Log file

https://gist.github.com/olimorris/dc262e730c720824701d34c9e01e3748

### PR DESCRIPTION
**Pull request by catlee**
Some LSP backends (`ruby-lsp` in particular here) can return a `nil` response to the `textDocument/diagnostic` method. This is contrary to the LSP spec, but unfortunately it means that neovim doesn't clear previous diagnostics when the `nil` response is received.

This PR changes the handling of the diagnostic response to treat a `nil` response the same as an empty `[]` response.

Fixes https://github.com/neovim/neovim/issues/26640

### CODE REVIEW

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (wookayin) [inline runtime/lua/vim/lsp/diagnostic.lua:309]: Following your logic, if `result == nil`, we'll still have `#diagnostics = 0` so this function would immediately return as it did previously?

**Comment by olimorris** [comment]
Just tested this PR out and can confirm that it **fixes** https://github.com/neovim/neovim/issues/26640. But I agree with @wookayin...this is an issue with Ruby-LSP. I'm going to raise an issue on there end too.

**Comment by catlee** [comment]
@wookayin are you suggesting that we still treat a nil response the same as a response with no items, with the addition of a warning to the user that their LSP backend is misbehaving?

**Comment by catlee** [comment]
I created https://github.com/Shopify/ruby-lsp/pull/1290 to get this fixed in `ruby-lsp`

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (MariaSolOs) [inline runtime/lua/vim/lsp/diagnostic.lua:311]: Nit but I find the `response` redundant.
```suggestion
        "LSP: client '%s' (id %s) returned nil to textDocument/diagnostic.\nThis is a bug in the LSP backend, and should be fixed within '%s'.",
```

**Review by MariaSolOs — APPROVED** [review]

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (wookayin) [inline runtime/lua/vim/lsp/diagnostic.lua:317]: While adding this ad-hoc warning message is fine in this PR, it should be noted that other handlers might throw errors (e.g. `workspace/applyEdit`) or silently ignore the invalid responses (e.g. `textDocument/definition`). (I'm not asking this should be fixed now) As we do further refactoring, this could be better handled in the common "boilerplate error validation and logging" wrapper for the handlers.

**Comment by wookayin** [comment]
Also please check the lintcommit CI output and update the title of PR and commits accordingly.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (catlee) [reply runtime/lua/vim/lsp/diagnostic.lua:302]: This PR does change the behaviour.

When result is nil, `diagnostics` (on line 301) will now be an empty table.

On line 320 we only exit early if there is no buffer attached for the file.

Otherwise we will continue down to the end of this function where on line 346 we call `vim.diagnostic.set` with an empty table, which clears the diagnostics.

I'll add a test for this.

Comment (wookayin) [reply runtime/lua/vim/lsp/diagnostic.lua:302]: I see, thanks for detailed explanation. `#diagnostics == 0` **and** `vim.fn.bufexists(fname) == 0`.

Actually I am not convinced that the handler should not behave so, when LSP responds an invalid response `nil`. To me, it feels like an over-correction specific to the behavior of the ruby LSP. What if it were meant to be "unchanged"? As a workaround, provided the LSP server is yet to be fixed, user *may* override the vim.lsp.handler to change the semantics of `nil` as "empty diagnostics".

Comment (catlee) [reply runtime/lua/vim/lsp/diagnostic.lua:302]: That's an excellent point. It's definitely possible that another possible interpretation of `nil` would be "unchanged".

What would it look like for the user to override the handler for a specific lsp to change the semantics for `nil`?

**Comment by gpanders** [comment]
I’m not in favor of adding workarounds for misbehaving servers, unless there is a *very* compelling reason. 

As a project we generally adhere to the position that we implement the specification *as-is*. If a server is non-compliant with the specification, it’s up to the server to fix that, not for Neovim to work around.

**Comment by wookayin** [comment]
(Moving the comment out of the review thread for better visibility)

> As a workaround, an user may override the handler to change the semantics of and behaviors on nil to "empty diagnostics", in case the LSP server is yet to be fixed and behaves buggy.\

> What would it look like for the user to override the handler for a specific lsp to change the semantics for nil?

An example is as follows:

```lua
---@param result lsp.DocumentDiagnosticReport
M["textDocument/diagnostic"] = function(err, result, ctx, config)
  if result == nil -- an invalid response, should never be nil (see #26957)
    -- 1. fail early
    -- error("invalid textDocument/diagnostic response, result is nil")

    -- 2. treat as if an empty response
    result = { kind = "full", items = {} } --[[ @as lsp.FullDocumentDiagnosticReport ]]

    -- 3. or treat as if an "unchanged" response
    -- result = { kind = "unchanged", resultId = "" } --[[ @as lsp.UnchangedDocumentDIagnosticReport ]]
  end
  return require('vim.lsp.diagnostic').on_diagnostic(err, result, ctx, config)
end
```

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (wookayin) [reply runtime/lua/vim/lsp/diagnostic.lua:302]: I've left comment for one example outside this thread.

**Comment by catlee** [comment]
> I’m not in favor of adding workarounds for misbehaving servers, unless there is a _very_ compelling reason.
> 
> As a project we generally adhere to the position that we implement the specification _as-is_. If a server is non-compliant with the specification, it’s up to the server to fix that, not for Neovim to work around.

I don't disagree in principle, but I'm also conscious of the user experience here.

VSCode handles `nil` responses like [this](https://github.com/microsoft/vscode-languageserver-node/blob/main/client/src/common/diagnostic.ts#L604):

```typescript
if (result === undefined || result === null || this.isDisposed || token.isCancellationRequested) {
  return { kind: vsdiag.DocumentDiagnosticReportKind.full, items: [] };
}
```

I'm assuming that the majority of LSP backend devs and the majority of devs using the LSP backends are using VSCode. As far as they're concerned, there's no bug here, everything is working fine.

**Comment by clason** [comment]
> I'm assuming that the majority of LSP backend devs and the majority of devs using the LSP backends are using VSCode. 

Yes, and that is precisely the problem. Nothing will change if someone doesn't hold Miss Feet to her words. So it's the specs, and nothing but the specs, ma'am.

**Comment by mfussenegger** [comment]
As others have already stated, we currently have a policy of no workarounds for servers that violate the spec and I don't see a reason to change that. If we'd relax that, we also make it harder for other editors and at some point the specification becomes less and less useful as each editor and server starts doing what they like and build up workarounds for each others quirks.

What I'd be okay with merging is an assert like `assert(response, "Server must return error or response, not nil")`.

**Comment by catlee** [comment]
> As others have already stated, we currently have a policy of no workarounds for servers that violate the spec and I don't see a reason to change that. If we'd relax that, we also make it harder for other editors and at some point the specification becomes less and less useful as each editor and server starts doing what they like and build up workarounds for each others quirks.

Ok, fair enough. I don't agree that this is the best thing for users in the short term, but understand why you want to keep to the spec.

> What I'd be okay with merging is an assert like `assert(response, "Server must return error or response, not nil")`.

I think the earlier suggestion of `notify_once` is more appropriate here, otherwise the user is spammed with error messages with a traceback every time the backend responds with `nil`.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (gpanders) [inline runtime/lua/vim/lsp/diagnostic.lua:305]: I would use "server" instead of "backend", as that's a more common term in LSP lingo.

Comment (gpanders) [inline runtime/lua/vim/lsp/diagnostic.lua:305]: I would use "server" instead of "backend", as that's a more common term in LSP lingo. 
```suggestion
        "LSP: client '%s' (id %s) returned nil to textDocument/diagnostic.\nThis is a bug in the LSP server, and should be fixed within '%s'.",
```

**Comment by MariaSolOs** [comment]
Not blocking for this PR, but I created https://github.com/microsoft/language-server-protocol/issues/1882 to make sure all LSP editors have a consistent behaviour around this.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (justinmk) [inline runtime/lua/vim/lsp/diagnostic.lua:308]: repeating the client name doesn't seem necessary

```suggestion
        "LSP: client '%s' (id %s) returned nil to textDocument/diagnostic.\nThis is a bug in the LSP server.",
        client_name,
        client_id
```

**Review by justinmk — APPROVED** [review]
LGTM after rebase (has merge conflicts)

**Pull request closed without merging**
